lib/widgets/autocomplete.dart

```diff
@@ -0,0 +1,174 @@
+  }
+
+  List<MentionAutocompleteResult> _resultsToDisplay = [];
+
+  void _viewModelChanged() {
+    setState(() {
+      _resultsToDisplay = _viewModel!.results.toList();
+    });
+  }
+
+  void _onTapOption(MentionAutocompleteResult option) {
+    // Probably the same intent that brought up the option that was tapped.
+    // If not, it still shouldn't be more out-of-date than the time it takes
+    // to compute the autocomplete results, which we do asynchronously.
+    final intent = widget.controller.autocompleteIntent();
+    if (intent == null) {
+      return; // Shrug.
+    }
+
+    final store = PerAccountStoreWidget.of(context);
+    final String replacementString;
+    switch (option) {
+      case UserMentionAutocompleteResult(:var userId):
+        // TODO(i18n) language-appropriate space character; check active keyboard?
+        //   (maybe handle centrally in `widget.controller`)
+        replacementString = '${mention(store.users[userId]!, silent: intent.query.silent)} ';
+      case WildcardMentionAutocompleteResult():
+        replacementString = '[unimplemented]'; // TODO
+      case UserGroupMentionAutocompleteResult():
+        replacementString = '[unimplemented]'; // TODO
+    }
+
+    widget.controller.value = intent.textEditingValue.replaced(
+      TextRange(
+        start: intent.syntaxStart,
+        end: intent.textEditingValue.selection.end),
+      replacementString,
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    return RawAutocomplete<MentionAutocompleteResult>(
+      textEditingController: widget.controller,
+      focusNode: widget.focusNode,
+      optionsBuilder: (_) => _resultsToDisplay,
+      // RawAutocomplete passes these when it calls optionsViewBuilder:
+      //   AutocompleteOnSelected<T> onSelected,
+      //   Iterable<T> options,
+      //
+      // We ignore them:
+      // - `onSelected` would cause some behavior we don't want,
+      //   such as moving the cursor to the end of the compose-input text.
+      // - `options` would be needed if we were delegating to RawAutocomplete
+      //   the work of creating the list of options. We're not; the
+      //   `optionsBuilder` we pass is just a function that returns
+      //   _resultsToDisplay, which is computed with lots of help from
+      //   MentionAutocompleteView.
+      optionsViewBuilder: (context, _, __) {
+        return Align(
+          alignment: Alignment.bottomLeft,
+          child: Material(
+            elevation: 4.0,
+            child: ConstrainedBox(
+              constraints: const BoxConstraints(maxHeight: 300), // TODO not hard-coded
+              child: ListView.builder(
+                padding: EdgeInsets.zero,
+                shrinkWrap: true,
+                itemCount: _resultsToDisplay.length,
+                itemBuilder: (BuildContext context, int index) {
+                  final option = _resultsToDisplay.elementAt(index);
```

Comment:
```suggestion
                  final option = _resultsToDisplay[index];
```

For lists they're equivalent; but `elementAt` is present on `Iterable`, so seeing it makes it look like we're potentially working with a general `Iterable` (where it might be slow).